doc/dev/mds_internals/locking.rst

```diff
@@ -0,0 +1,152 @@
+Ceph MDS Locker
+===============
+
+Why use locks?
+--------------
+
+Locking infrastructure in MDS is (obviously) to protect the state of various metadata. MDS has different locks covering different portions of inode and dentry. Moreover, MDS uses different kinds of locks since different metadata (in inode and dentry) have different behaviour in different situations. The MDS cache is distributed across multiple MDS ranks and across all clients. The locking infrastructure serves to ensure that all ranks and clients are consistent in their view of the file system.
+
+Data managed by the MDS can be very large to practically have the entire data set in the memory of a single metadata server. This also results in a single point of failure. The MDS therefore has a concept of Distributed Subtree Partition. A directory tree can be divided into smaller sub-trees. This is done by recording heat (access frequency) of each node in the directory tree. When a sub-tree heat reaches a configured threshold, the MDS divides the sub-tree by splitting the directory fragment. Each fragment is responsible for a part of the original directory, however, there will be a single authority node for these fragments. Each MDS can bear the read and write requests after fragmentation. If a file is very frequently accessed, the MDS will generate multiple copies distributed across active MDSs to satisfy concurrent I/Os. Typically, there are multiple clients reading and writing to files. The MDS defines locking rules for the associated metadata, e.g., metadata which is rarely modified concurrently such as UID/GID for an inode, a shared read and exclusive write access rule would suffice. However, statistics of a directory may need to be updated by multiple clients at the same time. This large directory may have been divided (fragmented) into multiple shards and different clients could write to different shards. These shards can share the read and also support simultaneous writes.
+
+Therefore, in addition to different lock types that cover different metadata peices for an inode, the MDS has lock classes that define access rules for a particular lock type. Lock types and classes are explained further in this document.
+
+Lock Types
+----------
+
+MDS defines a handful of lock types associated with different metadata for an inode or dentry. Lock type protecting metadata for an inode and dentry are as follows::
+
+  CEPH_LOCK_DN       - dentry
+  CEPH_LOCK_DVERSION - dentry version
+  CEPH_LOCK_IVERSION - inode version
+  CEPH_LOCK_IAUTH    - ctime, mode, uid, gid
+  CEPH_LOCK_ILINK    - ctime, nlink
+  CEPH_LOCK_IDFT     - dirfragtree, frags
+  CEPH_LOCK_IFILE    - ctime, mtime, atime, size, truncate_seq, truncate_size, client_ranges, inline_data
+  CEPH_LOCK_INEST    - rstats
+  CEPH_LOCK_IXATTR   - ctime, xattrs
+  CEPH_LOCK_ISNAP    - ctime, snaps
+  CEPH_LOCK_IFLOCK   - file locks
+  CEPH_LOCK_IPOLICY  - ctime, layout, quota, export_pin, ephemeral_*
```

Comment:
ctime is showing up a lot here and I'm not sure it should be. If this is correct, I think it needs more explaining.

Reply:
I think its correct. MDS updates ctime when inode attrs change - setting export_pin, etc...

Reply:
More explaining, then? You can't really be "protected" by more than one lock if they can be held at different times; ctime is just weird since the conflict resolution is "newest time" so it doesn't need much protecting.

Reply:
Hmmm... looks like ctime is either modified under `versionlock` or _after_ projecting an inode with other locks held. The recurrence of ctime (above) depicts that it _really_ does not need much protection.

I'll make that clear in this doc.